Scribe/MainEditorForm.cs

```diff
@@ -404,37 +406,58 @@ private void UpdateLibraryDataDisplay()
         {
+            // TODO Remove this debug statement.
+            MainToolStripStatusLabel.Text = ChangeManager.Count.ToString();
+
             if (sender is TextBox textbox
                 && string.Compare(textbox.Text,
                                   EditableControls[typeof(TextBox)][textbox],
                                   comparisonType: StringComparison.OrdinalIgnoreCase) != 0)
             {
-                EditableControls[typeof(TextBox)][textbox] = textbox.Text;
-                HasUnsavedChanges = true;
+                ChangeManager.AddAndExecute(new Change(EditableControls[typeof(TextBox)][textbox], textbox.Text, textbox.Name,
+                                            (object databaseValue) => { _ = databaseValue.ToString(); HasUnsavedChanges = true; },
+                                            (object displayValue) => textbox.Text = displayValue.ToString(),
+                                            (object oldValue) => EditableControls[typeof(TextBox)][textbox] = oldValue.ToString()));
+                
             }
             else if (sender is CheckBox checkbox
                      && string.Compare(checkbox.Checked.ToString(),
                                        EditableControls[typeof(CheckBox)][checkbox],
                                        comparisonType: StringComparison.OrdinalIgnoreCase) != 0)
             {
-                EditableControls[typeof(CheckBox)][checkbox] = checkbox.Checked.ToString();
-                HasUnsavedChanges = true;
+                var parsedOldValue = bool.TryParse(EditableControls[typeof(CheckBox)][checkbox], out var result)
+                    ? result
+                    : false;
+                ChangeManager.AddAndExecute(new Change(parsedOldValue, (bool?)checkbox.Checked, checkbox.Name,
+                                            (object databaseValue) => { _ = databaseValue; HasUnsavedChanges = true; },
+                                            (object displayValue) => checkbox.Checked = (bool)displayValue,
+                                            (object oldValue) => EditableControls[typeof(CheckBox)][checkbox] = oldValue.ToString()));
             }
             else if (sender is ComboBox combobox
                      && string.Compare(combobox.SelectedIndex.ToString(),
                                        EditableControls[typeof(ComboBox)][combobox],
                                        comparisonType: StringComparison.OrdinalIgnoreCase) != 0)
             {
-                EditableControls[typeof(ComboBox)][combobox] = combobox.SelectedIndex.ToString();
-                HasUnsavedChanges = true;
+                var parsedOldValue = int.TryParse(EditableControls[typeof(ComboBox)][combobox], out var result)
+                    ? result
+                    : 0;
+                ChangeManager.AddAndExecute(new Change(parsedOldValue, (int?)combobox.SelectedIndex, combobox.Name,
```

Comment:
You have int? here, and an assumed int (without ?) in one of the arguments. Should they be the same?

Reply:
Fixed!